src/lib/input/input-wrapper.ts

```diff
@@ -0,0 +1,283 @@
+  return [
+    'date',
+    'datetime',
+    'datetime-local',
+    'month',
+    'time',
+    'week'
+  ].filter(value => {
+    featureTestInput.setAttribute('type', value);
+    return featureTestInput.type === value;
+  });
+})();
+
+
+let nextUniqueId = 0;
+
+
+export class MdInputWrapperPlaceholderConflictError extends MdError {
+  constructor() {
+    super('Placeholder attribute and child element were both specified.');
+  }
+}
+
+
+export class MdInputWrapperUnsupportedTypeError extends MdError {
+  constructor(type: string) {
+    super(`Input type "${type}" isn't supported by md-input-wrapper.`);
+  }
+}
+
+
+export class MdInputWrapperDuplicatedHintError extends MdError {
+  constructor(align: string) {
+    super(`A hint was already declared for 'align="${align}"'.`);
+  }
+}
+
+
+/**
+ * The placeholder directive. The content can declare this to implement more
+ * complex placeholders.
+ */
+@Directive({
+  selector: 'md-placeholder'
+})
+export class MdPlaceholder {}
+
+
+/** The hint directive, used to tag content as hint labels (going under the input). */
+@Directive({
+  selector: 'md-hint',
+  host: {
+    'class': 'md-hint',
+    '[class.md-right]': 'align == "end"',
+  }
+})
+export class MdHint {
+  // Whether to align the hint label at the start or end of the line.
+  @Input() align: 'start' | 'end' = 'start';
+}
+
+
+/** The input directive, used to mark the input that `MdInputWrapper` is wrapping. */
+@Directive({
+  selector: 'input[md-input], textarea[md-input]',
+  host: {
+    'class': 'md-input-element',
+    '[id]': 'id',
+    '(blur)': '_onBlur()',
+    '(focus)': '_onFocus()',
+    '(input)': '_onInput()',
+  }
+})
+export class MdInputDirective implements AfterContentInit {
+  @Input()
+  get disabled() { return this._disabled; }
+  set disabled(value: any) { this._disabled = coerceBooleanProperty(value); }
+  private _disabled = false;
+
+  @Input()
+  get id() { return this._id; };
+  set id(value: string) { this._id = value || this._uid; }
+  private _id: string;
+
+  @Input()
+  get placeholder() { return this._placeholder; }
+  set placeholder(value: string) {
+    if (this._placeholder != value) {
+      this._placeholder = value;
+      this.placeholderChange.emit(this._placeholder);
+    }
+  }
+  private _placeholder = '';
+
+  @Input()
+  get required() { return this._required; }
+  set required(value: any) { this._required = coerceBooleanProperty(value); }
+  private _required = false;
+
+  @Input()
+  get type() { return this._type; }
+  set type(value: string) {
+    this._type = value || 'text';
+    this._validateType();
+  }
+  private _type = 'text';
+
+  @Input()
+  value: any;
+
+  @Output() placeholderChange = new EventEmitter<string>();
+
+  get empty() { return (this.value == null || this.value == '') && !this._isNeverEmpty(); }
+
+  focused = false;
+
+  private get _uid() { return this._cachedUid = this._cachedUid || `md-input-${nextUniqueId++}`; }
+  private _cachedUid: string;
```

Comment:
I was just trying to be extra defensive. I figured someone could do `[id]="myId"` and `myId` would initially have some value but it could theoretically update at some point in the future, so if it updates to `null` later on this will make sure we still put an id on it.